### PR TITLE
fix: backend validation does not prevent responses on hidden fields

### LINKED ISSUE #732: Backend validation does not prevent responses on hidden fields
Related to #7

Steps to reproduce:
- Create form with short text field hidden behind radio button
- Programatically submit short text field response without radio response. Submission is successful.

### LINKED ISSUE #779: Expand test cases to include legitimate storage mode submissions with hidden email/phone fields
> Upon further investigation, the team found that legitimate storage mode submissions containing responses to hidden email or phone number fields have been rejected - these are responses where the field was unhidden based on form logic. Due to end-to-end encryption, we are not able to identify if a storage mode field should be hidden or visible, and they retain their hidden state in the backend. This led to the deployed fix incorrectly blocking such submissions as responses were deleted on hidden fields.

As a follow-up to our postmortem, we should add test cases so that the above incident does not recur again.

### PR DESCRIPTION
**Pull request by tshuli**
## Problem
Builds on #736
Closes #732
Closes #779

## Solution
- Add check for isVisible in validateField() for email mode submissions
- Also adds / updates unit tests

## Tests
- [ ] Create a email mode form with Yes/No field, and a Number Field hidden behind it. Attempt to submit a number response programatically with the number field still hidden. Submission should fail.
- [ ] Now unhide the Number Field. Attempt to submit a response on the number field normally. Submission should pass.
- [ ] Create email mode form with all field types (including Yes/No) hidden behind a Yes/No field. Submit while keeping the fields hidden. Submission should pass.
- [ ] Create storage mode form with a mobile, email and short text field hidden behind a Yes/No field. Unhide the mobile and email fields and submit a response. Submission should pass.
- [ ] Create storage mode form with a mobile, email and short text field hidden behind a Yes/No field. Submit a response without unhiding the fields. Submission should pass.

### CODE REVIEW

**Comment by tshuli** [comment]
@mantariksh In this PR, I limited the scope of the hidden field response check to email mode forms only; see`/src/app/utils/field-validation/index.ts`:
```
  if (
    responseMode === ResponseMode.Email &&
    isResponsePresentOnHiddenField(response)
  ) {
    return err(
      new ValidateFieldError(`Attempted to submit response on a hidden field`),
    )
  }
```

We discussed that validateField is not appropriate to do logic validation, and we should instead set isVisible to be true for encrypt mode. This can be done in `src/app/modules/submission/submission.service.ts`:

```
    const processingResponse: ProcessedFieldResponse = {
      ...response,
      isVisible:
        // Set isVisible as true for Encrypt mode because we cannot tell
        // Prevents downstream validateField from incorrectly preventing
        // encrypt mode submissions with responses on hidden fields
        form.responseMode === ResponseMode.Encrypt
          ? true
          : visibleFieldIds.has(responseId),
      question: formField.getQuestion(),
    }
```

However for required fields, if the field is visible, we also check if the field is non-empty, for both email and encrypt mode forms - see `src/app/utils/field-validation/index.ts`. Setting isVisible to be true for all encrypted mode fields leads to an error here as there is no response for the required, 'visible' field. This means that we now instead have to check, in singleAnswerRequiresValidation(), if the form is an email or encrypt mode form.

```
const singleAnswerRequiresValidation = (
  formField: IField,
  response: ProcessedSingleAnswerResponse,
) => (formField.required && response.isVisible) || response.answer.trim() !== ''
```

Basically, there is no way around this issue because fundamentally both email and encrypt mode submissions are processed in the same way today . For the scope of this PR which is to implement a bug fix for responses in email mode submissions,  I suggest we stick to the more straightforward implementation which is to check for responses on hidden fields only on email mode submissions. When the sharding is done in #780, we can do away with isVisible for encrypt mode and this will solve the problem more fundamentally.

For your consideration

**Comment by tshuli** [comment]
@mantariksh as discussed, i've changed it to set isVisible = true for encrypt mode if there are responses. thanks for reviewing!

File: src/app/modules/submission/__tests__/submission/submission.service.spec.ts [hunk 1]
```diff
@@ -152,6 +152,43 @@ describe('submission.service', () => {
...
+      // Only check for mobile and email fields, since the other fields are
+      // e2e encrypted from the browser.
+      const mobileField = generateDefaultField(BasicField.Mobile)
+      const emailField = generateDefaultField(BasicField.Email)
+      // Add answers to both mobile and email fields
+      const mobileResponse = generateSingleAnswerResponse(
+        mobileField,
+        '+6587654321',
+      )
+      mobileResponse.isVisible = false
```
Comment (mantariksh) [inline src/app/modules/submission/__tests__/submission/submission.service.spec.ts:166]: this made me realise that the tests are no longer behaving as intended. the object which we pass into `getProcessedResponses` should be a `FieldResponse`, not some form of `ProcessedResponse`. might it be better to write a `generateFieldResponse` function which doesn't include `isVisible` or `isUserVerified`? or even better, rename `generateSingleAnswerResponse` to `generateProcessedSingleAnswerResponse`?

File: src/types/response/index.ts [hunk 2]
```diff
@@ -25,7 +25,7 @@ export interface ICheckboxResponse extends IBaseResponse {
   answerArray: string[]
 }
 
-export type ITableRow = string[]
+export type ITableRow = Array<string>
```
Comment (mantariksh) [inline src/types/response/index.ts:28]: nit: can we standardise to `string[]`? that's how it's written everywhere else in the codebase

File: tests/unit/backend/controllers/email-submissions.server.controller.spec.js [hunk 3]
```diff
@@ -1068,89 +1058,6 @@ describe('Email Submissions Controller', () => {
       prepareSubmissionThenCompare(expected, done)
     })
 
-    it('excludes field if isVisible is false for autoReplyData', (done) => {
```
Comment (mantariksh) [inline tests/unit/backend/controllers/email-submissions.server.controller.spec.js:1071]: why are all these tests deleted?

**Review by mantariksh — CHANGES_REQUESTED** [review]
main question is about deleted jasmine tests

File: src/app/utils/field-validation/index.ts [hunk 4]
```diff
@@ -45,6 +46,35 @@ const doFieldTypesMatch = (
...
 }
 
+/**
+ * Returns true if response appears on a hidden field.
+ * This may happen if a submission is made programatically to try and bypass form logic.
+ * @param response The submitted response
+ */
+const isResponsePresentOnHiddenField = (response: FieldResponse): boolean => {
+  if (isProcessedSingleAnswerResponse(response)) {
+    if (!response.isVisible && response.answer.trim() !== '') {
```
[the remark this answers is not in the document]
Comment (tshuli) [reply src/app/utils/field-validation/index.ts:59]: thanks! edited

File: src/app/modules/submission/__tests__/submission/submission.service.spec.ts (hunk 1 above)
Comment (tshuli) [reply src/app/modules/submission/__tests__/submission/submission.service.spec.ts]: edited

File: src/types/response/index.ts (hunk 2 above)
Comment (tshuli) [reply src/types/response/index.ts]: edited

File: tests/unit/backend/controllers/email-submissions.server.controller.spec.js (hunk 3 above)
Comment (tshuli) [reply tests/unit/backend/controllers/email-submissions.server.controller.spec.js]: Thanks for highlighting. I removed them initially as they were invalid since they were inputting responses to hidden fields, but on review, I've instead fixed the faulty values

File: src/app/utils/field-validation/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mantariksh) [reply src/app/utils/field-validation/index.ts:59]: ah but then you can also remove all the `!response.isVisible. subsequently!

File: tests/unit/backend/helpers/generate-form-data.ts [hunk 5]
```diff
@@ -117,12 +118,35 @@ export const generateSingleAnswerResponse = (
...
+      'Call the custom response generator functions for attachment, table and checkbox.',
+    )
+  }
+  return {
+    _id: field._id,
+    answer,
+    fieldType: field.fieldType as Exclude<
+      BasicField,
+      BasicField.Table | BasicField.Checkbox | BasicField.Attachment
+    >,
```
Comment (mantariksh) [inline tests/unit/backend/helpers/generate-form-data.ts:146]: this should include `question` right?

**Review by mantariksh — APPROVED** [review]
thank you for the comprehensive work on the tests!

File: tests/unit/backend/helpers/generate-form-data.ts (hunk 5 above)
Comment (tshuli) [reply tests/unit/backend/helpers/generate-form-data.ts:146]: nope, because question is only injected to ProcessedResponse; ISingleAnswerResponse (or FieldResponse in general) does not have question:

```
export interface IBaseResponse {
  _id: IFieldSchema['_id']
  fieldType: BasicField
  myInfo?: IMyInfo
  // Signature exists for verifiable fields if the answer is verified.
  signature?: string
}

export interface ISingleAnswerResponse extends IBaseResponse {
  fieldType: Exclude<
    BasicField,
    BasicField.Table | BasicField.Checkbox | BasicField.Attachment
  >
  answer: string
}
```

Comment (tshuli) [reply tests/unit/backend/helpers/generate-form-data.ts:146]: ^ thanks for highlighting, removed the unnecessary !response.isVisible
